cni/network/invoker_cns.go

```diff
@@ -356,3 +314,151 @@ func (invoker *CNSIPAMInvoker) Delete(address *net.IPNet, nwCfg *cni.NetworkConf
+		return errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
+	}
+
+	ncgw := net.ParseIP(info.ncGatewayIPAddress)
+	if ncgw == nil {
+		// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
+		if !overlayMode {
+			return errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+		}
+
+		if net.ParseIP(info.podIPAddress).To4() != nil { //nolint:gocritic
+			ncgw, err = getOverlayGateway(ncIPNet)
+			if err != nil {
+				return err
+			}
+		} else if net.ParseIP(info.podIPAddress).To16() != nil {
+			ncgw = net.ParseIP(overlayGatewayV6IP)
+		} else {
+			return errors.Wrap(err, "No podIPAddress is found: %w")
+		}
+	}
+
+	if ip := net.ParseIP(info.podIPAddress); ip != nil {
+		defaultInterfaceInfo := addResult.defaultInterfaceInfo.ipResult
+		defaultRouteDstPrefix := network.Ipv4DefaultRouteDstPrefix
+		if ip.To4() == nil {
+			defaultRouteDstPrefix = network.Ipv6DefaultRouteDstPrefix
+			addResult.ipv6Enabled = true
+		}
+
+		defaultInterfaceInfo.IPs = append(defaultInterfaceInfo.IPs,
+			&cniTypesCurr.IPConfig{
+				Address: net.IPNet{
+					IP:   ip,
+					Mask: ncIPNet.Mask,
+				},
+				Gateway: ncgw,
+			})
+
+		routes, getRoutesErr := getRoutes(info.routes, info.skipDefaultRoutes)
+		if getRoutesErr != nil {
+			return getRoutesErr
+		}
+
+		if len(routes) > 0 {
+			defaultInterfaceInfo.Routes = append(defaultInterfaceInfo.Routes, routes...)
+		} else { // add default routes if none are provided
+			defaultInterfaceInfo.Routes = append(defaultInterfaceInfo.Routes, &cniTypes.Route{
+				Dst: defaultRouteDstPrefix,
+				GW:  ncgw,
+			})
+		}
+
+		addResult.defaultInterfaceInfo.ipResult = defaultInterfaceInfo
+		addResult.defaultInterfaceInfo.skipDefaultRoutes = info.skipDefaultRoutes
+	}
+
+	// get the name of the primary IP address
+	_, hostIPNet, err := net.ParseCIDR(info.hostSubnet)
+	if err != nil {
+		return fmt.Errorf("unable to parse hostSubnet: %w", err)
+	}
+
+	addResult.hostSubnetPrefix = *hostIPNet
+	addResult.defaultInterfaceInfo.nicType = cns.InfraNIC
+
+	// set subnet prefix for host vm
+	// setHostOptions will execute if IPAM mode is not v4 overlay and not dualStackOverlay mode
+	// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
+	if !overlayMode {
+		if err := setHostOptions(ncIPNet, addConfig.options, info); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func configureSecondaryAddResult(info *IPResultInfo, addResult *IPAMAddResult, podIPConfig *cns.IPSubnet) error {
+	ip, ipnet, err := podIPConfig.GetIPNet()
+	if ip == nil {
+		return errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
+	}
+
+	macAddress, err := net.ParseMAC(info.macAddress)
+	if err != nil {
+		return errors.Wrap(err, "Invalid mac address")
+	}
+
+	result := InterfaceInfo{
+		ipResult: &cniTypesCurr.Result{
+			IPs: []*cniTypesCurr.IPConfig{
```

Comment:
didn't we agree changing this to []IPs, []Routes instead of using Result type

Reply:
[]Interfaces is still being used in parts of code, if we really need this refactor we should do in separate PR...

Reply:
discussed offline to not take dependency on cniTypesCurr.Result and will split in separate PR